### ISSUE TITLE
Ruby Library

### ISSUE THREAD

**Issue by ankane** [david-cortes/cmfrec#19]
Hi David, just wanted to let you know a [Ruby library](https://github.com/ankane/cmfrec) has been released. The ability to add side information and implicit features are both really great!

I was wondering if there's a way to get the same latent factors (`model.A_`, `model.B_`) between the Python and C libraries (this would make it easier to confirm everything is implemented correctly). I think the current difference may come from Python generating random values in the Cython code instead of using `reset_values` [here](https://github.com/david-cortes/cmfrec/blob/5b8d05ecc5c80707af8531f907532bda24a2358d/cmfrec/wrapper_untyped.pxi#L1158).

Also, I'm having trouble with the Windows DLL. It builds ([build code](https://github.com/ankane/cmfrec/blob/a98a78479ac900201aa19a45e24f766852aa4b5e/.github/workflows/build.yml) and [output](https://github.com/ankane/cmfrec/runs/1468323389?check_suite_focus=true)) but `dumpbin` reports no exported symbols. Not sure if you have any ideas here.

```txt
Dump of file vendor/cmfrec.dll

File Type: DLL

  Summary

        1000 .data
        1000 .pdata
        1000 .rdata
        1000 .reloc
        1000 .rsrc
        1000 .text
```

A few other features that would be neat are:

- similar items (item-item recommendations) and similar users for `CMF` and `CMF_implicit`
- validation set and early stopping options

All that said, great library! Thanks for building it!!

**Comment by david-cortes** [comment]
Thanks! Will add it to the readme.

About the other issues:
* I’m not sure that’s the right way to build in windows (`cmake --build .`). Do you get the same issue if you use `devenv` after cmake creates the build artifacts?
* I think finding similar users/items is beyond the scope of the library, as that’s better done with more specialized software (e.g. [hnswlib](https://github.com/nmslib/hnswlib)) – whatever simplistic procedure I could bundle in the package would most definitely not be competitive against such libraries or even against scikit-learn’s.
* Validation set and early stopping aren’t really needed when using ALS. You can try setting something like `niter=500` and you’ll see that you’ll actually get better results than with `niter=10` under every public dataset that you could think of. The only scenario in which it ends up getting worse with more iterations is if the regularization is too low. It could be a good idea for the gradient-based procedures though.

**Comment by david-cortes** [comment]
As for getting the same results in the factors from Python, it’s not possible as-is but you could edit the cython wrapper at the lines in which it calls `CMF` and `CMF_implicit`, setting `reset_values=true` and setting a random seed. It should give the same results as long as you use the same lapack library.

**Comment by david-cortes** [comment]
I tried using `MSBuild` and indeed it seems there was something wrong with the windows build. I've added an extra line to the cmake file but I'm not sure if it solves the issue. Could you try with the latest commit from master?

**Comment by ankane** [comment]
Thanks for adding it to the docs.

Re Windows: After d7106c59b558bcbed68d2dc5f1a258af3f86cbf9, symbols are exported, but now having trouble loading the DLL with FFI (`The specified module could not be found.`). Will dig a bit more on this (may be an issue on my end).

Re validation set/early stopping: That's great! Tried 500 iterations and confirmed better results.

Re similar items: Makes sense to use an ANN library. Just to confirm: only the latent factors are needed for this (`model.A_` for similar users and `model.B_` for similar items)? The bias (`model.user_bias_`, `model.item_bias_`), side information (`model.C_`, `model.D_`), and add implicit features (`model.Ai_`, `model.Bi_`) aren't needed/shouldn't be used?

Re `reset_values`: What do you think of having the Python and R packages use `reset_values` from C? Seems like it could reduce / DRY code while giving consistent results across packages.

**Comment by david-cortes** [comment]
For similar users/items: yes, only `A` and `B` are needed, optionally adding the biases as a last column.  For items I guess it’d be quite important to have the biases added. In theory I guess you could concatenate `Ai` and `Bi` as extra columns of `A` and `B` for similarity, but they are not used for prediction in any case (they are used to calculate `B` and `A`).

For `reset_values`, the issue is that this library was written in C, which has very poor random number generation capabilities out of the box. With `reset_values=true` it will call `larnv` from LAPACK, which uses an outdated algorithm with quite bad properties (among others: a very short RNG period by today’s standards), and tends to give worse result than using more modern methods such as MT or XOR-type algorithms. Those are available in e.g. NumPy, R, and C++>=11, but not in standard C.

**Comment by ankane** [comment]
Great, both make sense. Thanks again!

**Comment by david-cortes** [comment]
@ankane : I think I figured out the issue with the windows build this time. Turns out with MSVC it also needs a `.lib` file along with the `.dll`, which after the last commit should also be generated through the cmake script. Could you give it another try?

**Comment by ankane** [comment]
Hmm, still no luck. It should be able to be contained in a single `.dll` file (based on [other libraries](https://github.com/ankane/ml-builds)). Also, it looks like `CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS` isn't present in XGBoost's and LightGBM's CMake files, so there may be something else off.

**Comment by david-cortes** [comment]
I think the missing step for getting the dll to work in windows was to add `__declspec(dllimport)` in the header - should be fixed now.

**Comment by ankane** [comment]
Ah, that sounds right. Will try it out next time I'm working on cmfrec. Thanks @david-cortes!

**Comment by david-cortes** [comment]
> Thanks for adding it to the docs.
> 
> Re Windows: After [d7106c5](https://github.com/david-cortes/cmfrec/commit/d7106c59b558bcbed68d2dc5f1a258af3f86cbf9), symbols are exported, but now having trouble loading the DLL with FFI (`The specified module could not be found.`). Will dig a bit more on this (may be an issue on my end).
> 
> Re validation set/early stopping: That's great! Tried 500 iterations and confirmed better results.
> 
> Re similar items: Makes sense to use an ANN library. Just to confirm: only the latent factors are needed for this (`model.A_` for similar users and `model.B_` for similar items)? The bias (`model.user_bias_`, `model.item_bias_`), side information (`model.C_`, `model.D_`), and add implicit features (`model.Ai_`, `model.Bi_`) aren't needed/shouldn't be used?
> 
> Re `reset_values`: What do you think of having the Python and R packages use `reset_values` from C? Seems like it could reduce / DRY code while giving consistent results across packages.

@ankane Now starting from version 3.2.0, all versions use the same random number generator, so results should match exactly between Python-R-C, provided that the inputs are passed as sparse/dense matrices and that if there’s any side information, those matrices match in shape with the ratings matrix.

**Comment by ankane** [comment]
Awesome, that'll be really helpful in making sure everything is implemented correctly.
